### PR TITLE
done

### PR DESCRIPTION
**Pull request by artemrakov**
### Solutions for Lecture 3

Can you review my code, please ? :) 

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture3.hs [hunk 1]
```diff
@@ -82,7 +84,9 @@ Tuesday
   would work for **any** enumeration type in Haskell (e.g. 'Bool',
   'Ordering') and not just 'Weekday'?
 -}
-next = error "TODO"
+next :: (Enum a, Bounded a, Eq a) => a -> a
+next a | (maxBound `asTypeOf` a) == a = minBound `asTypeOf` a
```
Comment (chshersh) [inline src/Lecture3.hs]: Type inference works here so you don't need to use `asTypeOf` 🙂 
```suggestion
next a | maxBound == a = minBound
```

File: src/Lecture3.hs [hunk 2]
```diff
@@ -125,9 +135,14 @@ data Reward = Reward
     } deriving (Show, Eq)
 
 instance Semigroup Reward where
-
+  (<>) :: Reward -> Reward -> Reward
+  a <> b = Reward { rewardGold = foldMap rewardGold [a, b]
+                  , rewardSpecial = foldr ((||) . rewardSpecial) False [a, b]
```
Comment (chshersh) [inline src/Lecture3.hs]: If you want to write in the list-style, you can use the standard function `all` here 🙂 
```suggestion
                  , rewardSpecial = all rewardSpecial [a, b]
```

File: src/Lecture3.hs [hunk 3]
```diff
@@ -137,10 +152,12 @@ data List1 a = List1 a [a]
 
 -- | This should be list append.
 instance Semigroup (List1 a) where
+  (<>) :: List1 a -> List1 a -> List1 a
+  List1 x xs <> List1 y ys = List1 x (xs ++ (y : ys))
 
 
 {- | Does 'List1' have the 'Monoid' instance? If no then why?
-
+ No it cannot have it. As mempty method should works withour arguments and List1 requires argument in order to be created. As it cannot be empty
```
Comment (chshersh) [inline src/Lecture3.hs:160]: Absolutely correct 🆗

File: src/Lecture3.hs [hunk 4]
```diff
@@ -181,7 +205,8 @@ together only different elements.
 Product {getProduct = 6}
 
 -}
-appendDiff3 = error "TODO"
+appendDiff3 :: (Monoid a, Eq a) => a -> a -> a -> a
+appendDiff3 x y z = fold $ nub [x, y, z]
```
Comment (chshersh) [inline src/Lecture3.hs]: That's a clever and concise implementation! 👏🏻 

However, the type signature is more strict than required. Your function works with `Monoid`. But not every type can have a `Monoid` instance. This means, that `appendDiff3` can't work with types like `List1`. But if you change the constraint from `Monoid` to `Semigroup`, this function will now work. Though, relaxing the constraint will require changing the implementation as well 🚧

File: src/Lecture3.hs [hunk 5]
```diff
@@ -213,8 +238,15 @@ types that can have such an instance.
 -- instance Foldable Weekday where
 -- instance Foldable Gold where
 -- instance Foldable Reward where
--- instance Foldable List1 where
--- instance Foldable Treasure where
+instance Foldable List1 where
+  foldr :: (a -> b -> b) -> b -> List1 a -> b
+  foldr f acc (List1 x []) = f x acc
+  foldr f acc (List1 x (y:ys)) = foldr f (f y acc) (List1 x ys)
```
Comment (chshersh) [inline src/Lecture3.hs]: You don't need to implement a manual recursion here, you can reuse the `Foldable` instance for lists 📂

**Review by chshersh — APPROVED** [review]
Wow, exercises to Lecture 3 are done! 🏆 
You've almost finished the course 🤯 

Keep up the great work and I hope you're enjoying the course so far 🤗 
The finish line is near 🏁

File: src/Lecture3.hs (hunk 2 above)
Comment (artemrakov) [reply src/Lecture3.hs]: I think it should be `any`

**Comment by artemrakov** [comment]
Thanks for all the comments. I added all the fixed and merging pr. 🎉

**Pull request merged**